sparse/dok.py

```diff
@@ -0,0 +1,312 @@
+
+from numbers import Integral
+from collections import Iterable
+
+from .slicing import normalize_index
+from .utils import _zero_of_dtype
+
+try:  # Windows compatibility
+    int = long
+except NameError:
+    pass
+
+
+class DOK:
+    """
+    A class for building sparse multidimensional arrays.
+
+    Attributes
+    ----------
+    dtype : numpy.dtype
+        The datatype of this array. Can be :code:`None` if no elements
+        have been set yet.
+    shape : tuple[int]
+        The shape of this array.
+    dict : dict
+        The keys of this dictionary contain all the indices and the values
+        contain the nonzero entries.
+
+    See Also
+    --------
+    COO : A read-only sparse array.
+
+    Examples
+    --------
+    You can create :obj:`DOK` objects from Numpy arrays.
+
+    >>> x = np.eye(5, dtype=np.uint8)
+    >>> x[2, 3] = 5
+    >>> s = DOK(x)
+    >>> s
+    <DOK: shape=(5, 5), dtype=uint8, nnz=6>
+
+    You can also create them from just shapes, and use slicing assignment.
+
+    >>> s2 = DOK((5, 5))
+    >>> s2[1:3, 1:3] = [[4, 5], [6, 7]]
+    >>> s2
+    <DOK: shape=(5, 5), dtype=int64, nnz=4>
+
+    You can convert :obj:`DOK` arrays to :obj:`COO` arrays, or :obj:`numpy.ndarray`
+    objects.
+
+    >>> from sparse import COO
+    >>> s3 = COO(s2)
+    >>> s3
+    <COO: shape=(5, 5), dtype=int64, nnz=4, sorted=False, duplicates=False>
+    >>> s2.todense()  # doctest: +NORMALIZE_WHITESPACE
+    array([[0, 0, 0, 0, 0],
+           [0, 4, 5, 0, 0],
+           [0, 6, 7, 0, 0],
+           [0, 0, 0, 0, 0],
+           [0, 0, 0, 0, 0]])
+    """
+    def __init__(self, shape, values=None, dtype=None):
+        """
+        Create a :obj:`DOK` array from a Numpy array, :obj:`COO`, or a dict.
+
+        Parameters
+        ----------
+        shape : tuple[int]
+            The shape of the created array.
+        values : dict
+            The key-value pairs of indices and their corresponding data.
+        dtype : np.dtype
+            The data type of the array.
+
+        Examples
+        --------
+        You can create :obj:`DOK` objects from Numpy arrays or :obj:`COO` arrays.
+
+        >>> x = np.eye(5, dtype=np.uint8)
+        >>> x[2, 3] = 5
+        >>> s = DOK(x)
+        >>> s
+        <DOK: shape=(5, 5), dtype=uint8, nnz=6>
+
+        >>> from sparse import COO
+        >>> s2 = COO.from_numpy(np.eye(4, dtype=np.uint8))
+        >>> s2
+        <COO: shape=(4, 4), dtype=uint8, nnz=4, sorted=True, duplicates=False>
+        >>> s3 = DOK(s2)
+        >>> s3
+        <DOK: shape=(4, 4), dtype=uint8, nnz=4>
+
+        You can also create :obj:`DOK` arrays from a shape and a dict of
+        values. Zeros are automatically ignored.
+
+        >>> values = {
+        ...     (1, 2, 3): 4,
+        ...     (3, 2, 1): 0,
+        ... }
+        >>> s4 = DOK((5, 5, 5), values)
+        >>> s4
+        <DOK: shape=(5, 5, 5), dtype=int64, nnz=1>
+        """
+        from .coo import COO
+        self.dict = {}
```

Comment:
It seems odd to me to use the name of a core collection as a variable name.  I'm not sure I have a better solution though.  Maybe `.data`?

Reply:
`.data` sounds good to me. I thought of that at first, but wanted to make sure users don't confuse it with `COO.data`

Reply:
Happy with either.  I don't have strong thoughts here.